### PR TITLE
fix(core/http): fallback to github /gitlab hosttype

### LINKED ISSUE #11591: Regression with 26.15.0: GitLab dependencies no longer properly updated
### How are you running Renovate?

Self-hosted

### Please select which platform you are using if self-hosting.

gitlab.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

Latest version provided via renovate-bot/renovate-runner

### Describe the bug

Last week we started noticing that some GitLab dependencies managed using the `gitlabci-include` manager stopped getting updated when running the Renovate job. Looking at the logs we found some debug messages which looked suspicious (see below).

With older versions of Renovate (specifically 26.14.1) we hadn't seen these `GitLab API 404` and `Datasource 404` messages and even though they are logged at the DEBUG level, we get the impression that they hint at the problem with updating the `gitlab-include` dependency in question. We have now downgraded to 26.14.1 (using the renovate-runner v3.13.3) and everything appears to work again.

Everything else runs smoothly. Does this possibly have anything to do with the new feature implemented by https://github.com/renovatebot/renovate/pull/11226? Is it possible that the GitLab PAT token is missing in the request?

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: GitLab API 404 (repository=diemobiliar/it/glu/test/glu-test-shipment-service)
       "url": "https://gitlab.com/api/v4/projects/diemobiliar%2Fit%2Fglu%2Fservice%2Fglu-service-stack/repository/tags?per_page=100"
DEBUG: Datasource 404 (repository=diemobiliar/it/glu/test/glu-test-shipment-service)
       "datasource": "gitlab-tags",
       "lookupName": "diemobiliar/it/glu/service/glu-service-stack",
       "url": "https://gitlab.com/api/v4/projects/diemobiliar%2Fit%2Fglu%2Fservice%2Fglu-service-stack/repository/tags?per_page=100"
DEBUG: Failed to look up dependency diemobiliar/it/glu/service/glu-service-stack (repository=diemobiliar/it/glu/test/glu-test-shipment-service, packageFile=.gitlab-ci.yml, dependency=diemobiliar/it/glu/service/glu-service-stack)
```
and further down (part of a longer message starting with `DEBUG: packageFiles with updates`):
```
         "gitlabci-include": [
           {
             "packageFile": ".gitlab-ci.yml",
             "deps": [
               {
                 "datasource": "gitlab-tags",
                 "depName": "diemobiliar/it/glu/service/glu-service-stack",
                 "depType": "repository",
                 "currentValue": "1.0.2111",
                 "registryUrls": ["https://gitlab.com"],
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "diemobiliar/it/glu/service/glu-service-stack",
                     "message": "Failed to look up dependency diemobiliar/it/glu/service/glu-service-stack"
                   }
                 ],
                 "versioning": "semver"
               }
             ]
           }
         ],
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### LINKED ISSUE #11599: Private Go module support is broken since 26.7.0 (401 Unauthorized)
### How are you running Renovate?

Self-hosted

### Please select which platform you are using if self-hosting.

GitHub Enterprise Server

### If you're self-hosting Renovate, tell us what version of Renovate you run.

26.7.0

### Describe the bug

Since 26.7.0, pulling private go modules seems to be broken (it was working in 26.6.0).

Here is an [example repo](https://github.com/ocofaigh/go-private) where the issue can be reproduced: https://github.com/ocofaigh/go-private
In this example, there is a `go.mod` and `go.sum` file with both public and private modules defined. The private ones no longer work - renovate is giving a **401 Unauthorized**.

The runtime environment has the following set in order to be able to scan private modules (I don't think there is any problems here, as these work just fine with 26.6.0):
- `export GOPRIVATE=github.ibm.com`
- `git config --global url."https://${GITHUB_TOKEN}:x-oauth-basic@github.ibm.com/".insteadOf "https://github.ibm.com/"`
- `echo -e "machine github.ibm.com\n  login ${GITHUB_TOKEN}" > ~/.netrc`

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Matched 1 file(s) for manager gomod: go.mod (repository=OCOFAIGH/go-test)
DEBUG: Found gomod package files (repository=OCOFAIGH/go-test)
DEBUG: Found 1 package file(s) (repository=OCOFAIGH/go-test)
 INFO: Dependency extraction complete (repository=OCOFAIGH/go-test)
       "baseBranch": "master",
       "stats": {
         "managers": {"gomod": {"fileCount": 1, "depCount": 6}},
         "total": {"fileCount": 1, "depCount": 6}
       }
DEBUG: Go lookup import url (repository=OCOFAIGH/go-test)
       "goModule": "github.ibm.com/IAM/pep/v2",
       "goImportURL": "https://github.ibm.com/IAM/pep.git"
DEBUG: Go lookup import url (repository=OCOFAIGH/go-test)
       "goModule": "github.ibm.com/IAM/token/v3",
       "goImportURL": "https://github.ibm.com/IAM/token.git"
DEBUG: Unknown GitHub error (repository=OCOFAIGH/go-test)
       "err": {
         "name": "HTTPError",
         "timings": {
           "start": 1630933924102,
           "socket": 1630933924102,
           "lookup": 1630933924134,
           "connect": 1630933924232,
           "secureConnect": 1630933924336,
           "upload": 1630933924337,
           "response": 1630933924463,
           "end": 1630933924463,
           "phases": {
             "wait": 0,
             "dns": 32,
             "tcp": 98,
             "tls": 104,
             "request": 1,
             "firstByte": 126,
             "download": 0,
             "total": 361
           }
         },
         "message": "Response code 401 (Unauthorized)",
         "stack": "HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/26.7.0 (https://github.com/renovatebot/renovate)",
             "accept": "application/vnd.github.v3+json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://github.ibm.com/api/v3/repos/IAM/pep/tags?per_page=100",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         },
         "response": {
           "statusCode": 401,
           "statusMessage": "Unauthorized",
           "body": {
             "message": "Must authenticate to access this API.",
             "documentation_url": "https://docs.github.com/enterprise/2.22/rest"
           },
           "headers": {
             "server": "openresty",
             "date": "Mon, 06 Sep 2021 13:12:06 GMT",
             "content-type": "application/json; charset=utf-8",
             "content-length": "118",
             "connection": "close",
             "x-github-enterprise-version": "2.22.18",
             "x-github-media-type": "github.v3; format=json",
             "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset",
             "access-control-allow-origin": "*",
             "x-github-request-id": "1232a435-7d7d-4aed-986c-392342066cc2",
             "strict-transport-security": "max-age=31536000; includeSubdomains",
             "x-frame-options": "deny",
             "x-content-type-options": "nosniff",
             "x-xss-protection": "1; mode=block",
             "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
             "content-security-policy": "default-src 'none'",
             "x-runtime-rack": "0.027829"
           },
           "httpVersion": "1.1"
         }
       }
DEBUG: Datasource unauthorized (repository=OCOFAIGH/go-test)
       "datasource": "go",
       "lookupName": "github.ibm.com/IAM/pep/v2",
       "url": "https://github.ibm.com/api/v3/repos/IAM/pep/tags?per_page=100"
DEBUG: Failed to look up dependency github.ibm.com/IAM/pep/v2 (repository=OCOFAIGH/go-test, packageFile=go.mod, dependency=github.ibm.com/IAM/pep/v2)
DEBUG: Unknown GitHub error (repository=OCOFAIGH/go-test)
       "err": {
         "name": "HTTPError",
         "timings": {
           "start": 1630933924142,
           "socket": 1630933924143,
           "lookup": 1630933924146,
           "connect": 1630933924246,
           "secureConnect": 1630933924346,
           "upload": 1630933924346,
           "response": 1630933924483,
           "end": 1630933924484,
           "phases": {
             "wait": 1,
             "dns": 3,
             "tcp": 100,
             "tls": 100,
             "request": 0,
             "firstByte": 137,
             "download": 1,
             "total": 342
           }
         },
         "message": "Response code 401 (Unauthorized)",
         "stack": "HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/26.7.0 (https://github.com/renovatebot/renovate)",
             "accept": "application/vnd.github.v3+json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://github.ibm.com/api/v3/repos/IAM/token/tags?per_page=100",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         },
         "response": {
           "statusCode": 401,
           "statusMessage": "Unauthorized",
           "body": {
             "message": "Must authenticate to access this API.",
             "documentation_url": "https://docs.github.com/enterprise/2.22/rest"
           },
           "headers": {
             "server": "openresty",
             "date": "Mon, 06 Sep 2021 13:12:06 GMT",
             "content-type": "application/json; charset=utf-8",
             "content-length": "118",
             "connection": "close",
             "x-github-enterprise-version": "2.22.18",
             "x-github-media-type": "github.v3; format=json",
             "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset",
             "access-control-allow-origin": "*",
             "x-github-request-id": "36566766-7fec-402f-bdc7-0da47270eaf1",
             "strict-transport-security": "max-age=31536000; includeSubdomains",
             "x-frame-options": "deny",
             "x-content-type-options": "nosniff",
             "x-xss-protection": "1; mode=block",
             "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
             "content-security-policy": "default-src 'none'",
             "x-runtime-rack": "0.016562"
           },
           "httpVersion": "1.1"
         }
       }
DEBUG: Datasource unauthorized (repository=OCOFAIGH/go-test)
       "datasource": "go",
       "lookupName": "github.ibm.com/IAM/token/v3",
       "url": "https://github.ibm.com/api/v3/repos/IAM/token/tags?per_page=100"
DEBUG: Failed to look up dependency github.ibm.com/IAM/token/v3 (repository=OCOFAIGH/go-test, packageFile=go.mod, dependency=github.ibm.com/IAM/token/v3)
DEBUG: Go lookup import url (repository=OCOFAIGH/go-test)
       "goModule": "github.ibm.com/project-fortress/vault-client-golang",
       "goImportURL": "https://github.ibm.com/project-fortress/vault-client-golang.git"
DEBUG: Unknown GitHub error (repository=OCOFAIGH/go-test)
       "err": {
         "name": "HTTPError",
         "timings": {
           "start": 1630933924804,
           "socket": 1630933924805,
           "lookup": 1630933924806,
           "connect": 1630933924911,
           "secureConnect": 1630933925216,
           "upload": 1630933925216,
           "response": 1630933925334,
           "end": 1630933925335,
           "phases": {
             "wait": 1,
             "dns": 1,
             "tcp": 105,
             "tls": 305,
             "request": 0,
             "firstByte": 118,
             "download": 1,
             "total": 531
           }
         },
         "message": "Response code 401 (Unauthorized)",
         "stack": "HTTPError: Response code 401 (Unauthorized)\n    at Request.<anonymous> (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:117:42)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/26.7.0 (https://github.com/renovatebot/renovate)",
             "accept": "application/vnd.github.v3+json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://github.ibm.com/api/v3/repos/project-fortress/vault-client-golang/tags?per_page=100",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         },
         "response": {
           "statusCode": 401,
           "statusMessage": "Unauthorized",
           "body": {
             "message": "Must authenticate to access this API.",
             "documentation_url": "https://docs.github.com/enterprise/2.22/rest"
           },
           "headers": {
             "server": "openresty",
             "date": "Mon, 06 Sep 2021 13:12:07 GMT",
             "content-type": "application/json; charset=utf-8",
             "content-length": "118",
             "connection": "close",
             "x-github-enterprise-version": "2.22.18",
             "x-github-media-type": "github.v3; format=json",
             "access-control-expose-headers": "ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset",
             "access-control-allow-origin": "*",
             "x-github-request-id": "dba43079-5406-469d-b888-5089423ce10b",
             "strict-transport-security": "max-age=31536000; includeSubdomains",
             "x-frame-options": "deny",
             "x-content-type-options": "nosniff",
             "x-xss-protection": "1; mode=block",
             "referrer-policy": "origin-when-cross-origin, strict-origin-when-cross-origin",
             "content-security-policy": "default-src 'none'",
             "x-runtime-rack": "0.013887"
           },
           "httpVersion": "1.1"
         }
       }
DEBUG: Datasource unauthorized (repository=OCOFAIGH/go-test)
       "datasource": "go",
       "lookupName": "github.ibm.com/project-fortress/vault-client-golang",
       "url": "https://github.ibm.com/api/v3/repos/project-fortress/vault-client-golang/tags?per_page=100"
DEBUG: Failed to look up dependency github.ibm.com/project-fortress/vault-client-golang (repository=OCOFAIGH/go-test, packageFile=go.mod, dependency=github.ibm.com/project-fortress/vault-client-golang)
DEBUG: Package releases lookups complete (repository=OCOFAIGH/go-test)
       "baseBranch": "master"
```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by viceice**
## Changes:

We now search for `hostType=github` again, when we send `hostType=github-tags` and other platform relates hoste types.
Same fallback for gitlab.

## Context:
fixes #11599 #11591

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/platform/github/index.ts [hunk 1]
```diff
@@ -115,12 +115,6 @@ export async function initPlatform({
     renovateUsername,
   };
 
-  // Generic github hostRule that per default all datasources using github api are enabled
-  const genericGithubHostRule = {
-    matchHost: parseUrl(defaults.endpoint).hostname,
-    token,
-  };
-  hostRules.add(genericGithubHostRule);
```
Comment (viceice) [inline lib/platform/github/index.ts:123]: We already addind one here:
https://github.com/renovatebot/renovate/blob/fb75b2f4dc8e72bc5457800c386310ad020da403/lib/platform/index.ts#L58-L70

File: lib/util/http/host-rules.spec.ts [hunk 2]
```diff
@@ -111,4 +116,40 @@ describe('util/http/host-rules', () => {
...
+
+  it('noAuth', () => {
+    expect(applyHostRules(url, { ...options, noAuth: true }))
+      .toMatchInlineSnapshot(`
+      Object {
+        "hostType": "github",
+        "noAuth": true,
+      }
+    `);
+  });
```
Comment (viceice) [inline lib/util/http/host-rules.spec.ts:128]: 100% coverage for `lib/utils/http` 😉

File: lib/util/http/index.spec.ts [hunk 3]
```diff
@@ -208,4 +208,11 @@ describe('util/http/index', () => {
     expect(bar).toBeTrue();
     expect(baz).toBeTrue();
   });
+
+  it('getBuffer', async () => {
+    httpMock.scope(baseUrl).get('/').reply(200, Buffer.from('test'));
+    const res = await http.getBuffer('http://renovate.com');
+    expect(res.body).toBeInstanceOf(Buffer);
+    expect(res.body.toString('utf-8')).toEqual('test');
+  });
```
Comment (viceice) [inline lib/util/http/index.spec.ts:217]: coverage

File: lib/util/http/index.ts [hunk 4]
```diff
@@ -53,12 +53,16 @@ export interface HttpResponse<T = string> {
...
+): HttpResponse<T> {
+  const { body, statusCode, headers } = response;
   // clone body and headers so that the cached result doesn't get accidentally mutated
+  // Don't use json clone for buffers
   return {
-    statusCode: response.statusCode,
-    body: clone<T>(response.body),
-    headers: clone(response.headers),
+    statusCode,
+    body: body instanceof Buffer ? (body.slice() as T) : clone<T>(body),
```
Comment (viceice) [inline lib/util/http/index.ts:64]: Do not use json serialization for buffer, we loose type

File: lib/util/http/index.ts [hunk 5]
```diff
@@ -166,6 +171,7 @@ export class Http<GetOptions = HttpOptions, PostOptions = HttpPostOptions> {
       resPromise = memCache.get(cacheKey);
     }
 
+    // istanbul ignore else: no chache tests
```
Comment (viceice) [inline lib/util/http/index.ts:174]:
```suggestion
    // istanbul ignore else: no cache tests
```
Ups

File: lib/util/http/auth.ts [hunk 6]
```diff
@@ -1,9 +1,11 @@
...
-import { PLATFORM_TYPE_GITEA } from '../../constants/platforms';
-import { GITHUB_API_USING_HOST_TYPES } from '../../types';
-import { GITLAB_API_USING_HOST_TYPES } from '../../types/platform/gitlab';
-import { GotOptions } from './types';
+import type { NormalizedOptions } from 'got';
+import {
+  GITHUB_API_USING_HOST_TYPES,
+  GITLAB_API_USING_HOST_TYPES,
+  PLATFORM_TYPE_GITEA,
+} from '../../constants/platforms';
```
Comment (viceice) [inline lib/util/http/auth.ts:7]: Just moved constants

**Review by rarkins — APPROVED** [review]

**Pull request merged**